library/java/org/chromium/net/impl/CronetBidirectionalStream.java

```diff
@@ -25,157 +36,93 @@
 
-  @GuardedBy("mNativeStreamLock")
-  // Metrics information. Obtained when request succeeds, fails or is canceled.
-  private RequestFinishedInfo.Metrics mMetrics;
-
-  /* Native BidirectionalStream object, owned by CronetBidirectionalStream. */
-  @GuardedBy("mNativeStreamLock") private long mNativeStream;
-
-  /**
-   * Read state is tracking reading flow.
-   *                         / <--- READING <--- \
-   *                         |                   |
-   *                         \                   /
-   * NOT_STARTED -> STARTED --> WAITING_FOR_READ -> READING_DONE -> SUCCESS
-   */
-  @GuardedBy("mNativeStreamLock") private @State int mReadState = State.NOT_STARTED;
-
-  /**
-   * Write state is tracking writing flow.
-   *                         / <---  WRITING  <--- \
-   *                         |                     |
-   *                         \                     /
-   * NOT_STARTED -> STARTED --> WAITING_FOR_FLUSH -> WRITING_DONE -> SUCCESS
-   */
-  @GuardedBy("mNativeStreamLock") private @State int mWriteState = State.NOT_STARTED;
+  private WriteBuffer mLastWriteBufferSent;
+  private ByteBuffer mLatestBufferRead;
+  private int mLatestBufferReadInitialPosition;
+  private int mLatestBufferReadInitialLimit;
 
   // Only modified on the network thread.
   private UrlResponseInfoImpl mResponseInfo;
 
-  /*
-   * OnReadCompleted callback is repeatedly invoked when each read is completed, so it
-   * is cached as a member variable.
-   */
-  // Only modified on the network thread.
-  private OnReadCompletedRunnable mOnReadCompletedTask;
-
   private Runnable mOnDestroyedCallbackForTesting;
 
   private final class OnReadCompletedRunnable implements Runnable {
     // Buffer passed back from current invocation of onReadCompleted.
-    ByteBuffer mByteBuffer;
+    private ByteBuffer mByteBuffer;
     // End of stream flag from current invocation of onReadCompleted.
-    boolean mEndOfStream;
+    private final boolean mEndOfStream;
+
+    OnReadCompletedRunnable(ByteBuffer mByteBuffer, boolean mEndOfStream) {
+      this.mByteBuffer = mByteBuffer;
+      this.mEndOfStream = mEndOfStream;
+    }
 
     @Override
     public void run() {
       try {
         // Null out mByteBuffer, to pass buffer ownership to callback or release if done.
         ByteBuffer buffer = mByteBuffer;
         mByteBuffer = null;
-        boolean maybeOnSucceeded = false;
-        synchronized (mNativeStreamLock) {
-          if (isDoneLocked()) {
-            return;
-          }
-          if (mEndOfStream) {
-            mReadState = State.READING_DONE;
-            maybeOnSucceeded = (mWriteState == State.WRITING_DONE);
-          } else {
-            mReadState = State.WAITING_FOR_READ;
-          }
+        @NextAction
+        int nextAction =
+            mState.nextAction(mEndOfStream ? Event.LAST_READ_COMPLETED : Event.READ_COMPLETED);
+        if (nextAction == NextAction.INVOKE_ON_READ_COMPLETED_CALLBACK) {
```

Comment:
Commenting here but this is more of a general question: is `mState.nextAction(@Event)` good enough to really know the next action or do you also need to know the context within it is being called (i.e., does an action have different meanings in different contexts)?

I'm trying to understand if it would be possible for me to reason about this as a simple `hashmap<@Action, functionExecutinAction>` or if there are more nuances (I would love to be able to reason about the next action logic and the action to function mapping on their own).

Reply:
Many NextAction are found in more than one place in `CronetBidirectionalStream`. What is a little bit better are the Events: they usually present in only one method (there are exception that I can fix).

The Mapping is a bit more complicated: 

```
Map<Pair<Event, State>, NextAction>
```

As far as I can tell, there are only two action that are generic: CARRY_ON and TAKE_NO_MORE_ACTIONS. So it is not impossible to rather return a Function, but it will require some extra consideration for these two generic NextAction.